paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/shuffle/RangeShuffle.java

```diff
@@ -154,20 +156,35 @@ public static <T> DataStream<Tuple2<T, RowData>> rangeShuffleByKey(
                                         BasicTypeInfo.INT_TYPE_INFO, input.getOutputType()),
                                 input.getParallelism());
 
-        // 4. Remove the partition id. (shuffle according range partition)
-        return new DataStream<>(
-                inputDataStream.getExecutionEnvironment(),
-                new OneInputTransformation<>(
-                        new PartitionTransformation<>(
-                                preparePartition,
-                                new CustomPartitionerWrapper<>(
-                                        new AssignRangeIndexOperator.RangePartitioner(rangeNum),
-                                        new AssignRangeIndexOperator.Tuple2KeySelector<>()),
-                                StreamExchangeMode.BATCH),
-                        "REMOVE KEY",
-                        new RemoveRangeIndexOperator<>(),
-                        input.getOutputType(),
-                        outParallelism));
+        PartitionTransformation<Tuple2<Integer, Tuple2<T, RowData>>> customPartitioner =
+                new PartitionTransformation<>(
+                        preparePartition,
+                        new CustomPartitionerWrapper<>(
+                                new AssignRangeIndexOperator.RangePartitioner(rangeNum),
+                                new AssignRangeIndexOperator.Tuple2KeySelector<>()),
+                        StreamExchangeMode.BATCH);
+
+        // 4. Remove the range index or both range index and key. (shuffle according range
+        // partition)
+        if (outputWithKey) {
```

Comment:
You don't need to change here.
You can just output with key, and introduce a RemoveKeyOperator.

It should have no impact to performance if it is chaining to upstream operator.

Reply:
The outputWithKey parameter is indeed unnecessary, and the changes have been made as requested.